rasa/graph_components/adders/nlu_prediction_to_history_adder.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+import logging
+from rasa.shared.core.events import UserUttered
+from typing import Dict, Text, Any, List
+
+from rasa.core.channels.channel import UserMessage
+
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.nlu.constants import ENTITIES, INTENT, TEXT
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.core.trackers import DialogueStateTracker
+from rasa.shared.core.domain import Domain
+
+logger = logging.getLogger(__name__)
+
+
+class NLUPredictionToHistoryAdder(GraphComponent):
+    """Adds NLU predictions to DialogueStateTracker."""
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> NLUPredictionToHistoryAdder:
+        """Creates component (see parent class for full docstring)."""
+        return cls()
+
+    def add(
+        self,
+        predictions: List[Message],
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        original_message: UserMessage,
+    ) -> DialogueStateTracker:
+        """Adds NLU predictions to the tracker.
+
+        Args:
+            predictions: A list of NLU predictions wrapped as Messages
+            tracker: The tracker the predictions should be attached to
+            domain: The domain of the model.
+            original_message: An original message from the user with
+                extra metadata to annotate the predictions (e.g. channel)
+
+        Returns:
+            The original tracker updated with events created from the predictions
+        """
+        for message in predictions:
+            user_utterance = UserUttered(
+                message.data.get(TEXT),
+                message.data.get(INTENT),
+                message.data.get(ENTITIES),
+                message.data,
```

Comment:
I actually think it would be best to leave this out because`UserUttered.parse_data` would get assigned [here](https://github.com/RasaHQ/rasa/blob/e152bba0f5c28ccb21041d2bcb8beb9786c00d48/rasa/shared/core/events.py#L450) with the correct metadata from the provided `original_message`, otherwise it might get overwritten [here](https://github.com/RasaHQ/rasa/blob/e152bba0f5c28ccb21041d2bcb8beb9786c00d48/rasa/shared/core/events.py#L461) if metadata in `message.data` is `None`?

Reply:
That's a good spot - I wasn't sure if the `parse_data` was going to contain other info that needed to be preserved which is why I did it that way but that's a good point. I'll change it out since we're trying to preserve the `metadata` on the `original_message` deliberately so that's probably the priority